File: combination-sum-iv.java
class Solution {    
    private int[] dp;
    
    public int combinationSum4(int[] nums, int target) {
        dp = new int[target + 1];
        Arrays.fill(dp, -1);
        return helper(nums, target);
    }
    
    private int helper(int[] nums, int target) {
        if (target == 0) {
            return 1;
        }
        
        if (target < 0) {
            return 0;
        }
        
        if (dp[target] != -1) {
            return dp[target];
        }
        
        int count = 0;
        for(int num : nums) {
            count += helper(nums, target - num);
        }
        
        dp[target] = count;
        return count;
    }
}

File: combination-sum-iv/combination-sum-iv.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        return self.helper(nums, target)
    
    def helper(self, nums, target):
        if target < 0:
            return 0
        
        dp = [0] * (target + 1)
        dp[0] = 1
        for i in range(1, target + 1):
            for num in nums:
                if i >= num:
                    dp[i] += dp[i - num]
        
        return dp[target]